018-cli-2.0/018-cli-2.0.md

```diff
@@ -0,0 +1,312 @@
+
+  ![CLI Auth](cli_auth.png)
+
+  This approach may seem unnecessarily complex now since we only allow email password based logins in the console. But as we move towards supporting OAuth logins in the console, this is the approach that needs to be followed. [Firebase CLI](https://github.com/firebase/firebase-tools/blob/0b0459bfde3fabf41223bb6d5d39a5cf325f1010/src/auth.ts#L476) for reference. 
+
+
+## Easier Deployment of cloud functions
+--- 
+
+### 🟢 Constrain deployments from current directory.
+
+We start by removing the `--code` parameter in the `createTag` command and constrain function deployments only from the current directory. 
+
+The refactored command will look like this 
+
+```sh
+appwrite functions createTag --functionId=[ID] --command=[COMMAND]
+```
+
+This corresponding appwrite issue will also need to be tackled as we implement this.
+https://github.com/appwrite/appwrite/issues/1316
+
+### 🟢 The `.appwrite.json` File
+
+The next improvement is going to be along the lines of a `.appwrite.json` file. Running `appwrite init` in the current directory will create a `.appwrite.json` file and initialize the current directory with attributes necessary to communicate with a particular project.
+
+Consider this example
+```sh
+mkdir temp && cd temp
+
+# Initialize the current directory with an Appwrite project 
+appwrite init
+
+temp 
+└── .appwrite.json
+```
+
+A `.appwrite.json` file will be associated with one project only whose ID will be visible in the `project` attribute. 
+
+```json
+{
+    "project" : "",
+    "functions": [
+        {
+          "name" : "My awesome function 💪",
+          "path" : "./FunctionOne",
+          "command" : "python main.py",
+          "runtime": "python3.9",
+          // This section is useful when we allow cloud functions to be tested locally
+          "environment_variables" :{
+            "KEY 1" : "VALUE 1",
+          }
+        },
+    ],
+    "hosting": {}
+}
+```
+
+There are two scenarios here 
+* A developer would like to link the directory to an existing project
+
+```sh
+appwrite init
+
+Choose a project to link this directory to 
+⭕️ Project 1
+✅️ Project 2
+⭕️ Project 3
+```
+
+* A developer would like to create a new project while initializing this directory
+
+```sh
+appwrite init --new
+
+Give your project a name:
+Project X
+
+Creating Project X...
+
+Project X Created 👍
+```
+
+> We need to first add support to utopia-php/CLI to allow for single select and multi select commands to enable this feature.
+
+
+### 🟢 The `appwrite deploy` command
+
+The `appwrite deploy` command is a convenient wrapper around three main commands
+* Create Function
+* Create Tag
+* Update Tag
+
+```sh
+# Usage 1
+appwrite deploy functions
+Which functions would you like to deploy? 
+✅️ FunctionOne
+✅️ FunctionTwo
+⭕️ FunctionThree
+
+# Usage 2
+appwrite deploy functions --all
+# Deploys all the functions specified in .appwrite.json
+
+# Future support for static hosting
+appwrite deploy site
```

Comment:
We should choose a consistent name in the future (site/hosting).

Reply:
Yes sure :+1: